unconstrained_test.go

```diff
@@ -884,16 +915,286 @@ func (Linear) Df(x, grad []float64) {
+		x:       []float64{-1.2, 1},
+		optVal:  0,
+		optLoc:  []float64{1, 1},
+		gradTol: defaultGradientAbsTol,
+	},
+	{
+		f:      Rosenbrock{},
+		x:      []float64{-1.2, 1},
+		optVal: 0,
+		optLoc: []float64{1, 1},
+	},
+	{
+		f:       Rosenbrock{},
+		x:       []float64{-120, 100, 50},
+		optVal:  0,
+		optLoc:  []float64{1, 1, 1},
+		gradTol: defaultGradientAbsTol,
+	},
+	{
+		f:      RosenbrockFDf{},
+		x:      []float64{-1.2, 1},
+		optVal: 0,
+		optLoc: []float64{1, 1},
+	},
+	{
+		f:       RosenbrockFDf{},
+		x:       []float64{-120, 100, 50},
+		optVal:  0,
+		optLoc:  []float64{1, 1, 1},
+		gradTol: defaultGradientAbsTol,
+	},
+	{
+		f:      HelicalValley{},
+		x:      []float64{-1, 0, 0},
+		optVal: 0,
+		optLoc: []float64{1, 0, 0},
+	},
+	{
+		f:      BiggsEXP4{},
+		x:      []float64{1, 2, 1, 1},
+		optVal: 0,
+		optLoc: []float64{1, 10, 1, 5},
+	},
+}
+
+var newtonTests = []UnconstrainedTest{
+	{
+		f:      Rosenbrock{},
+		x:      []float64{-12000, 10000},
+		optVal: 0,
+		optLoc: []float64{1, 1},
+	},
+	{
+		f:       Rosenbrock{},
+		x:       []float64{10, 10, 10, 10},
+		optVal:  0,
+		optLoc:  []float64{1, 1, 1, 1},
+		gradTol: 1e-11,
+	},
+	{
+		f:      BiggsEXP6{},
+		x:      []float64{1, 2, 1, 1, 1, 1},
+		optVal: 0.005655649925,
+	},
+	{
+		f:       Gaussian{},
+		x:       []float64{0.4, 1, 0},
+		optVal:  1.12793e-8,
+		optLoc:  []float64{0.3989561, 1.0000191, 0},
+		gradTol: 1e-10,
+	},
+	{
+		f:       Powell{},
+		x:       []float64{0, 1},
+		optVal:  0,
+		optLoc:  []float64{1.09815933e-5, 9.10614674},
+		gradTol: 1e-9,
+	},
+	{
+		f:      Box{},
+		x:      []float64{0, 10, 20},
+		optVal: 0,
+		optLoc: []float64{1, 10, 1},
+	},
+	generateVariablyDimensioned(10, 1e-13),
+	generateVariablyDimensioned(100, 1e-10),
+	{
+		f:       Watson{},
+		x:       []float64{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+		optVal:  0,
+		optLoc:  []float64{1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1},
+		gradTol: 1e-8,
+	},
+	{
+		f:      Penalty1{},
+		x:      []float64{1, 2, 3, 4},
+		optVal: 2.2499775e-5,
+	},
+	{
+		f:      Penalty1{},
+		x:      []float64{1, 2, 3, 4, 5, 6, 7, 8, 9, 10},
+		optVal: 7.0876515e-5,
+	},
+	{
+		f:      Penalty2{},
+		x:      []float64{0.5, 0.5, 0.5, 0.5},
+		optVal: 9.37629e-6,
+	},
+	{
+		f:      Penalty2{},
+		x:      []float64{0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5},
+		optVal: 2.93660e-4,
+	},
+	{
+		f:       BrownDennis{},
+		x:       []float64{25, 5, -5, -1},
+		optVal:  85822.20162635628,
+		gradTol: 1e-10,
+	},
+	{
+		f:      GulfRD{},
+		x:      []float64{5, 2.5, 0.15},
+		optVal: 0,
+		optLoc: []float64{50, 25, 1.5},
+	},
+	{
+		f:      ExtendedPowell{},
+		x:      []float64{3, -1, 0, 1},
+		optVal: 0,
+		optLoc: []float64{0, 0, 0, 0},
+	},
+	{
+		f:      Beale{},
+		x:      []float64{1, 1},
+		optVal: 0,
+		optLoc: []float64{3, 0.5},
+	},
+	{
+		f:      Wood{},
+		x:      []float64{-3, -1, -3, -1},
+		optVal: 0,
+		optLoc: []float64{1, 1, 1, 1},
+	},
+}
+
+var bfgsTests = []UnconstrainedTest{
```

Comment:
There are problems that BFGS and LBFGS handle equally well, but then there are problems that one method handles better than the other (e.g., can achieve lower gradient norm, converges from a point much further from the minimum, ...). The idea was to put the former problems into newtonTests and the former into bfgsTests or lbfgsTests - if a method can solve a more difficult problem, then why not include it in the tests. I was not sure about it either and I hoped that it would come up during the review.

This touches a little on the issue of the purpose of these tests - are we testing for correctness of the implementation, for the ability to solve as hard problems as possible (badly scaled, stringent tolerances, ...) or some other criteria not yet included?